src/analysis/vector/geometry_checker/qgsgeometrygapcheck.cpp

```diff
@@ -19,27 +19,69 @@
 #include "qgsgeometrycollection.h"
 #include "qgsfeaturepool.h"
 #include "qgsvectorlayer.h"
+#include "qgsvectorlayerutils.h"
 #include "qgsfeedback.h"
 #include "qgsapplication.h"
+#include "qgsproject.h"
 
 #include "geos_c.h"
 
 QgsGeometryGapCheck::QgsGeometryGapCheck( const QgsGeometryCheckContext *context, const QVariantMap &configuration )
   : QgsGeometryCheck( context, configuration )
   ,  mGapThresholdMapUnits( configuration.value( QStringLiteral( "gapThreshold" ) ).toDouble() )
+{
+}
 
+void QgsGeometryGapCheck::prepare( const QgsGeometryCheckContext *context, const QVariantMap &configuration )
 {
+  if ( configuration.value( QStringLiteral( "allowedGapsEnabled" ) ).toBool() )
+  {
+    QgsVectorLayer *layer = context->project->mapLayer<QgsVectorLayer *>( configuration.value( "allowedGapsLayer" ).toString() );
+    mAllowedGapsLayer = layer;
+    mAllowedGapsSource = qgis::make_unique<QgsVectorLayerFeatureSource>( layer );
 
+    mAllowedGapsBuffer = configuration.value( QStringLiteral( "allowedGapsBuffer" ) ).toDouble();
+  }
+  else
+  {
+    mAllowedGapsSource.reset();
+  }
 }
 
 void QgsGeometryGapCheck::collectErrors( const QMap<QString, QgsFeaturePool *> &featurePools, QList<QgsGeometryCheckError *> &errors, QStringList &messages, QgsFeedback *feedback, const LayerFeatureIds &ids ) const
 {
   if ( feedback )
     feedback->setProgress( feedback->progress() + 1.0 );
 
-  QVector<QgsGeometry> geomList;
+  std::unique_ptr<QgsAbstractGeometry> allowedGapsGeom;
+  std::unique_ptr<QgsGeometryEngine> allowedGapsGeomEngine;
 
+  if ( mAllowedGapsSource )
+  {
+    QVector<QgsGeometry> allowedGaps;
+    QgsFeatureRequest request;
+    request.setSubsetOfAttributes( QgsAttributeList() );
+    QgsFeatureIterator iterator = mAllowedGapsSource->getFeatures( request );
+    QgsFeature feature;
 
+    while ( iterator.nextFeature( feature ) )
+    {
+      QgsGeometry geom = feature.geometry();
+      QgsGeometry gg = geom.buffer( mAllowedGapsBuffer, 20 );
+      allowedGaps.append( gg );
+    }
+
+    std::unique_ptr< QgsGeometryEngine > allowedGapsEngine = QgsGeometryCheckerUtils::createGeomEngine( nullptr, mContext->tolerance );
+    allowedGapsEngine->prepareGeometry();
```

Comment:
No need for this -- prepared geometries are used in GEOS combine operations (only the spatial predicates utilise prepared geometries)